sys/net/application_layer/gcoap/forward_proxy/forward_proxy.c

```diff
@@ -0,0 +1,232 @@
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Cenk Gündoğan <cenk.guendogan@haw-hamburg.de>
+ */
+
+#include "net/gcoap.h"
+#include "net/gcoap/forward_proxy.h"
+#include "uri_parser.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static uint8_t proxy_req_buf[CONFIG_GCOAP_PDU_BUF_SIZE];
+static uint8_t scratch[48];
+
+static bool _parse_endpoint(sock_udp_ep_t *remote,
+                            uri_parser_result_t *urip)
+{
+    ipv6_addr_t addr;
+    remote->family = AF_INET6;
+
+    /* copy host to scratch for safe string operations with '\0' */
+    if (urip->host_len >= sizeof(scratch)) {
+        return false;
+    }
+    memcpy(scratch, urip->host, urip->host_len);
+    scratch[urip->host_len] = '\0';
+
+    char *addr_str = (char *) scratch;
+    uint16_t addr_str_len = urip->host_len;
+
+    /* support IPv6 only for now */
+    if (addr_str[0] != '[') {
+        return false;
+    }
+
+    /* check for interface */
+    char *iface_str = strchr(addr_str, '%');
+    if (iface_str) {
+        /* do not count '%' and ']' */
+        unsigned iface_len = addr_str_len - (iface_str - addr_str) - 2;
+
+        /* also do not count '%' */
+        addr_str_len -= iface_len + 1;
+
+        /* skip '%' */
+        iface_str++;
+        iface_str[iface_len] = '\0';
+
+        int pid = atoi(iface_str);
+        if (gnrc_netif_get_by_pid(pid) == NULL) {
+            return false;
+        }
+        remote->netif = pid;
+    }
+    /* no interface present in host string */
+    else {
+        if (gnrc_netif_numof() == 1) {
+            /* assign the single interface found in gnrc_netif_numof() */
+            remote->netif = (uint16_t)gnrc_netif_iter(NULL)->pid;
+        }
+        else {
+            remote->netif = SOCK_ADDR_ANY_NETIF;
+        }
+    }
+
+    /* skip '[' and reduce length by 2 for '[' and ']' */
+    addr_str++;
+    addr_str_len -= 2;
+
+    /* replace ']' with '\0' for safe string operations */
+    addr_str[addr_str_len] = '\0';
+
+    /* parse destination address */
+    if (ipv6_addr_from_str(&addr, addr_str) == NULL) {
+        return false;
+    }
+    if ((remote->netif == SOCK_ADDR_ANY_NETIF) && ipv6_addr_is_link_local(&addr)) {
+        return false;
+    }
+    memcpy(&remote->addr.ipv6[0], &addr.u8[0], sizeof(addr.u8));
+
+    /* copy port string into scratch for atoi */
+    memcpy(scratch, urip->port, urip->port_len);
+    scratch[urip->port_len] = '\0';
+
+    remote->port = atoi((char *) scratch);
+
+    if (remote->port == 0) {
+        return false;
+    }
+
+    return true;
+}
+
+static void _forward_resp_handler(const gcoap_request_memo_t *memo,
+                                  coap_pkt_t* pdu,
+                                  const sock_udp_ep_t *remote)
+{
+    (void) remote; /* this is the origin server */
+
+    /* forward the response packet as-is to the client */
+    gcoap_dispatch((uint8_t *)pdu->hdr,
```

Comment:
There will need to be some compensation for mismatching token lengths (see the other comment about the empty token).